csharp/ql/src/semmle/code/csharp/frameworks/JsonNET.qll

```diff
@@ -0,0 +1,233 @@
+  class JsonClass extends Class { JsonClass() { this.getParent() instanceof JsonNETNamespace } }
+
+  /** The class `Newtonsoft.Json.JsonConvert`. */
+  class JsonConvertClass extends JsonClass, LibraryTypeDataFlow {
+    JsonConvertClass() { this.hasName("JsonConvert") }
+
+    /** Gets a `ToString` method. */
+    private Method getAToStringMethod() {
+      result = this.getAMethod("ToString") and
+      result.isStatic()
+    }
+
+    /** Gets a `Deserialize` method. */
+    Method getADeserializeMethod() {
+      result = this.getAMethod() and
+      result.getName().matches("Deserialize%")
+    }
+
+    /** Gets a `Serialize` method. */
+    Method getASerializeMethod() {
+      result = this.getAMethod() and
+      result.getName().matches("Serialize%")
+    }
+
+    private Method getAPopulateMethod() {
+      result = this.getAMethod() and
+      result.getName().matches("Populate%")
+    }
+
+    override predicate callableFlow(
+      CallableFlowSource source, CallableFlowSink sink, SourceDeclarationCallable c,
+      boolean preservesValue
+    ) {
+      // ToString methods
+      c = getAToStringMethod() and
+      preservesValue = true and
+      source = any(CallableFlowSourceArg arg | arg.getCallable() = c and arg.getArgumentIndex() = 0) and
+      sink instanceof CallableFlowSinkReturn
+      or
+      // Deserialize methods
+      c = getADeserializeMethod() and
+      preservesValue = false and
+      source = any(CallableFlowSourceArg arg | arg.getCallable() = c and arg.getArgumentIndex() = 0) and
+      sink instanceof CallableFlowSinkReturn
+      or
+      // Serialize methods
+      c = getASerializeMethod() and
+      preservesValue = false and
+      source = any(CallableFlowSourceArg arg | arg.getCallable() = c and arg.getArgumentIndex() = 0) and
+      sink instanceof CallableFlowSinkReturn
+      or
+      // Populate methods
+      c = getAPopulateMethod() and
+      preservesValue = false and
+      source = any(CallableFlowSourceArg arg | arg.getCallable() = c and arg.getArgumentIndex() = 0) and
+      sink = any(CallableFlowSinkArg arg | arg.getCallable() = c and arg.getArgumentIndex() = 1)
+    }
+  }
+
+  /** A type that is serialized. */
+  private class SerializedType extends ValueOrRefType {
+    SerializedType() {
+      // Supplied as a parameter to a serialize or deserialize method
+      exists(TypeParameter tp, JsonConvertClass jc, UnboundGenericMethod serializeMethod |
+        this = tp.getAnUltimatelySuppliedType() and
+        tp = serializeMethod.getATypeParameter()
+      |
+        serializeMethod = jc.getASerializeMethod()
+        or
+        serializeMethod = jc.getADeserializeMethod()
+      )
+      or
+      exists(Class attribute | attribute = this.getAnAttribute().getType() |
+        attribute instanceof JsonObjectAttributeClass
+        or
+        attribute.hasName("JsonConverterAttribute")
+        or
+        attribute.hasQualifiedName("System.DataContractAttribute")
```

Comment:
So Json.NET does respect these attributes, but I agree it's not specific to JSON.net. I've just removed it as I doubt very much it will cause many false-negatives.